framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -189,6 +193,52 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
         return new RegularTransferResult<>(newResultValue, result.getRegularStore());
     }
 
+    private TransferResult<CFValue, CFStore> createNewConditionalResult(
+            TransferResult<CFValue, CFStore> result, List<Boolean> resultValues, Node node) {
+        AnnotationMirror stringVal = createBooleanAnnotationMirror(resultValues);
+        CFValue newResultValue =
+                analysis.createSingleAnnotationValue(
+                        stringVal, result.getResultValue().getUnderlyingType());
+
+        return new ConditionalTransferResult<>(
+                newResultValue, result.getThenStore(), result.getElseStore(), true);
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitFieldAccess(
+            FieldAccessNode node, TransferInput<CFValue, CFStore> in) {
+
+        TransferResult<CFValue, CFStore> result = super.visitFieldAccess(node, in);
+
+        // If array.length is encountered, transform its @IntVal annotation into an @ArrayLen annotation for array.
+        if (NodeUtils.isArrayLengthFieldAccess(node)) {
+            CFValue value =
+                    result.getRegularStore()
+                            .getValue(
+                                    FlowExpressions.internalReprOf(
+                                            analysis.getTypeFactory(), node));
+            if (value != null) {
+                AnnotationMirror lengthAnno =
+                        AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntVal.class);
+                if (lengthAnno != null) {
+                    List<Long> lengthValues = ValueAnnotatedTypeFactory.getIntValues(lengthAnno);
+                    List<Integer> arrayLenValues = new ArrayList<>(lengthValues.size());
+                    for (Long l : lengthValues) {
+                        arrayLenValues.add(l.intValue());
+                    }
+                    AnnotationMirror newArrayAnno =
+                            atypefactory.createArrayLenAnnotation(arrayLenValues);
+                    Receiver arrayRec =
+                            FlowExpressions.internalReprOf(
+                                    analysis.getTypeFactory(), node.getReceiver());
+                    result.getRegularStore().clearValue(arrayRec);
```

Comment:
You should glb with the previous value rather than clearing it. See test case I added in framework/tests/value/Refinement2.java.